### PR TITLE
[JENKINS-36084] Responsive new job name validation

### PR DESCRIPTION
**Pull request by samatdav**
Currently the GET request to check new job name for invalid characters is sent only on 'blur' jQuery event.
That makes a user know about the invalid character only after he entered the name and forces to go back to input field.

This change checks for invalid characters in new job name after each character (keyup event).
Additionally, it removes unnecessary warning messages when job name is empty.

### CODE REVIEW

**Comment by michaelneale** [comment]
looks like jshint error now: 

`03:06:32 [ERROR] 
03:06:32 [ERROR] events.js:141
03:06:32 [ERROR]       throw er; // Unhandled 'error' event
03:06:32 [ERROR]       ^
03:06:32 [ERROR] Error: JSHint failed for: src/main/js/add-item.js
`

**Comment by samatdav** [comment]
@michaelneale can you tell how to approach this JSHint error?
Also I don't understand why the remote check fails, but shows that all tests passed: https://jenkins.ci.cloudbees.com/job/core/job/jenkins-core/5484/testReport/

**Comment by oleg-nenashev** [comment]
Is it WiP?

**Comment by samatdav** [comment]
@oleg-nenashev I fixed the issued I wanted to solve and now do not see problems. So now I plan to ask mentors (and may be others) if they have other comments and if the change looks good to them.

**Comment by samatdav** [comment]
@lanwen @recena do you have comments on the change? What it needs to be approved?

**Comment by recena** [comment]
@samatdav I'd like to review this PR.

**Comment by recena** [comment]
@samatdav If you are working on [JENKINS-36084](https://issues.jenkins-ci.org/browse/JENKINS-36084) it is helpful if you keep the JIRA issue status updated.

**Comment by oleg-nenashev** [comment]
Modified the PR status according to the feedback

File: war/src/main/js/add-item.js [hunk 1]
```diff
@@ -246,15 +245,23 @@ $.when(getItems()).done(function(data) {
             showInputHelp('.add-item-name');
             setFieldValidationStatus('name', true);
             if (getFormValidationStatus()) {
-              enableSubmit(true);
+              if (event.type === 'blur') {
```
Comment (lanwen) [inline war/src/main/js/add-item.js]: maybe it will be look better with switch-case

File: war/src/main/js/add-item.js [hunk 2]
```diff
@@ -246,15 +245,23 @@ $.when(getItems()).done(function(data) {
             showInputHelp('.add-item-name');
             setFieldValidationStatus('name', true);
             if (getFormValidationStatus()) {
-              enableSubmit(true);
+              if (event.type === 'blur') { 
+                enableSubmit(true);
+              } else if (event.type === 'keyup') {
+                var btn = $('form .footer .btn-decorator button[type=submit]');
```
Comment (lanwen) [inline war/src/main/js/add-item.js]: what does it mean?

Comment (samatdav) [inline war/src/main/js/add-item.js]: Since enable submit contains btn.focus() in it, that is to disable the focus of the button for keyup event. So when event was keyup, only necessary part of enable submit is called.

Comment (samatdav) [inline war/src/main/js/add-item.js]: The btn variable defined again since it was defined only locally before and needed again now.

Comment (lanwen) [inline war/src/main/js/add-item.js]: can u reuse some code to define this action? It looks now like a temp solution

Comment (daniel-beck) [inline war/src/main/js/add-item.js]: Would it be possible to separate enabling and focusing, and re-use enabling here?

**Comment by oleg-nenashev** [comment]
ping @samatdav

File: war/src/main/js/add-item.js [hunk 3]
```diff
@@ -246,15 +257,22 @@ $.when(getItems()).done(function(data) {
             showInputHelp('.add-item-name');
             setFieldValidationStatus('name', true);
             if (getFormValidationStatus()) {
-              enableSubmit(true);
+              switch (event.type) {
+                case 'blur':
+                  enableSubmit(true);
+                  break;
+                case 'keyup':
+                  enableButton(true);
```
Comment (samatdav) [inline war/src/main/js/add-item.js]: @daniel-beck @lanwen Is this how you suggested to separate focusing part of enableSubmit feature?

Comment (daniel-beck) [inline war/src/main/js/add-item.js]: @samatdav Pretty much, yes.

**Comment by batmat** [comment]
Also ping @jenkinsci/code-reviewers

**Comment by daniel-beck** [comment]
One weird interaction. This stupid form still has autocompletion support, so any time I enter a job name similar to an existing job name, I get the stupid dropdown from my browser that hides the validation area.

Previously, this wasn't a big deal, as validation only kicked in after leaving the input field. Now, validation is shown immediately, but hidden by the dropdown.

This could (maybe even should) be addressed in the context of this PR to make this change actually useful -- as I expect name duplication to be more common than using terrible characters, and there's an overlap here.

What do others think?

**Comment by michaelneale** [comment]
@daniel-beck as in it pops up suggestions for new names based on existing names? that seems really out of place here. The only time that could be useful would be if you are establishing a naming convention and it makes it easy to stick to things, otherwise shouldn't that be turned off?

**Comment by daniel-beck** [comment]
@michaelneale 

> otherwise shouldn't that be turned off? 

It's a web browser feature. This is one of the few form fields in Jenkins that don't have this disabled, and the problem is that the dropdown overlaps the validation error.

I.e. while this PR appears to accomplish everything it set out to do, it's not working great due to the browser's form field autocompletion feature that gets into the way

**Comment by michaelneale** [comment]
@samatdav it seems that `autocomplete=off` needs to be an attribute set somewhere as per https://developer.mozilla.org/en-US/docs/Web/Security/Securing_your_site/Turning_off_form_autocompletion for that name field (I don't think people would object)

oh dear: 

> In some case, the browser will keep suggesting autocompletion values even if the autocomplete attribute is set to off. This unexpected behavior can be quite puzzling for developers. The trick to really force the no-completion is to assign a random string to the attribute like so :
> 
> autocomplete="nope"
>

**Comment by samatdav** [comment]
@tfennelly @recena Could you provide feedback on both the proposed change and the autocomplete issue (in last comments)? Do you suggest to change anything?

File: war/src/main/js/add-item.js [hunk 4]
```diff
@@ -116,6 +116,18 @@ $.when(getItems()).done(function(data) {
       }
     }
 
+    function focusButton(status) {
+      var btn = $('form .footer .btn-decorator button[type=submit]');
+      if (status === true) {
+        btn.focus();
+      }
```
Comment (tfennelly) [inline war/src/main/js/add-item.js]: Should there be an `else` here to remove focus?

Comment (tfennelly) [inline war/src/main/js/add-item.js]: Also ... as things are, there's no need to do the jquery select until inside the `if` block i.e. at the moment it's redundant what `status === false`.

File: war/src/main/js/add-item.js [hunk 5]
```diff
@@ -116,6 +116,18 @@ $.when(getItems()).done(function(data) {
       }
     }
 
+    function focusButton(status) {
+      var btn = $('form .footer .btn-decorator button[type=submit]');
```
Comment (tfennelly) [inline war/src/main/js/add-item.js]: I think add a `getFormSubmitButton` function that hides the jquery selector and can be reused here and in the `enableButton` function.

File: war/src/main/js/add-item.js [hunk 6]
```diff
@@ -100,14 +100,14 @@ $.when(getItems()).done(function(data) {
...
       var btn = $('form .footer .btn-decorator button[type=submit]');
       if (status === true) {
         if (btn.hasClass('disabled')) {
           btn.removeClass('disabled');
           btn.prop('disabled', false);
         }
-        btn.focus();
       } else {
         if (!btn.hasClass('disabled')) {
           btn.addClass('disabled');
```
Comment (tfennelly) [inline war/src/main/js/add-item.js]: I know the above code is not part of this PR, but it could be cleaned up. The `btn.hasClass('disabled')` calls are dumb. I'm sure the above code could be more simply stated as ...

```javascript
function enableButton(status) {
  var btn = $('form .footer .btn-decorator button[type=submit]');

  btn.prop('disabled', !status);
  if (status === true) {
    btn.removeClass('disabled');
  } else {
    btn.addClass('disabled');
  }
}
```

File: war/src/main/js/add-item.js (hunk 4 above)
Comment (samatdav) [inline war/src/main/js/add-item.js]: `else` functionality was not present initially and seems that there may not be much need for it. Or do you think it would be better with `else` ?

Ok, I moved selecting inside if block.

File: war/src/main/js/add-item.js (hunk 6 above)
Comment (samatdav) [inline war/src/main/js/add-item.js]: Thanks, I added your code.

File: war/src/main/js/add-item.js (hunk 5 above)
Comment (samatdav) [inline war/src/main/js/add-item.js]: I got an error trying to use 
```
function getFormSubmitButton() {
    var btn = $('form .footer .btn-decorator button[type=submit]');
}
```
and then just calling it inside.
Do you think it is necessary to replace it?

**Comment by samatdav** [comment]
@tfennelly I implemented most of your suggestions. Do you have any other objections? Do you have an opinion about autocomplete?

File: war/src/main/js/add-item.js (hunk 5 above)
Comment (tfennelly) [inline war/src/main/js/add-item.js]: Try ...

```javascript
function getFormSubmitButton() {
    return $('form .footer .btn-decorator button[type=submit]');
}
```

And then to use it...

```javascript
var btn = getFormSubmitButton();

// etc...
```

**Comment by tfennelly** [comment]
@samatdav wrt autocomplete ... assuming there'd be objections to removing it, how about moving the validation warning so as the autocomplete doesn't hide it?

**Comment by samatdav** [comment]
@tfennelly I am afraid that will create even more debate on whether and how the validation warning should be moved.
I plan to make a PR with UI changes but I am afraid it will take forever to discuss. 
Here are some of my UI ideas. May be a part of it would be considered viable.
https://github.com/samatdav/jenkins/pull/2/files

**Comment by samatdav** [comment]
@tfennelly In https://github.com/jenkinsci/jenkins/pull/2501 I propose to move input box so that autocomplete will not hide warning message. What do you think?

**Comment by lanwen** [comment]
How it looks like now?

**Comment by samatdav** [comment]
@lanwen 
out-3

**Comment by daniel-beck** [comment]
@tfennelly WDYT? Can't help it, this looks weird…

**Comment by samatdav** [comment]
@daniel-beck @lanwen @michaelneale I removed the shift and added a red border. What do you think?
out-4

**Comment by lanwen** [comment]
red border looks much better

**Comment by daniel-beck** [comment]
The suggestion to indicate an invalid title through styling of the text field came up in today's meeting about Samat's GSoC project. I like it as well, but wonder whether a background color could make it even more noticeable without distracting so much (you want people to notice, after all!)

That said, even if it's just the border, I'm now 👍 on this. Further refinements can be done later, and Samat got the live validation done.

File: war/src/main/js/add-item.js [hunk 7]
```diff
@@ -234,27 +240,36 @@ $.when(getItems()).done(function(data) {
...
       if (!isItemNameEmpty()) {
         var itemName = $('input[name="name"]', '#createItem').val();
         $.get("checkJobName", { value: itemName }).done(function(data) {
           var message = parseResponseFromCheckJobName(data);
           if (message !== '') {
             activateValidationMessage('#itemname-invalid', '.add-item-name', message);
+            $('#add-item-panel input').css('border', "1px solid red");
           } else {
             cleanValidationMessages('.add-item-name');
+            $('#add-item-panel input').css('border', "1px solid #999");
```
Comment (daniel-beck) [inline war/src/main/js/add-item.js]: Can't this use CSS classes that already exist?

Comment (samatdav) [inline war/src/main/js/add-item.js]: @daniel-beck Is it better know? 
https://github.com/jenkinsci/jenkins/pull/2416/files#diff-e146d4fdbfe6385a456aea9775f6282dR250

Comment (samatdav) [inline war/src/main/js/add-item.js]: red

File: war/src/main/js/add-item.js [hunk 8]
```diff
@@ -234,27 +240,36 @@ $.when(getItems()).done(function(data) {
...
     // Init NameField
-    $('input[name="name"]', '#createItem').blur(function() {
+    $('input[name="name"]', '#createItem').on('keyup blur', function(event) {
       if (!isItemNameEmpty()) {
         var itemName = $('input[name="name"]', '#createItem').val();
         $.get("checkJobName", { value: itemName }).done(function(data) {
           var message = parseResponseFromCheckJobName(data);
           if (message !== '') {
             activateValidationMessage('#itemname-invalid', '.add-item-name', message);
+            $('input[name="name"]').css({'border': "1px solid red", "background-color": "#FADEDE"});
```
Comment (lanwen) [inline war/src/main/js/add-item.js]: don't add exact css, toggle class

**Comment by samatdav** [comment]
@lanwen I added class toggle. Is it fine now?

File: war/src/main/js/widgets/add/addform.less [hunk 9]
```diff
@@ -14,6 +14,11 @@
     z-index: 5;
   }
 
+  .warning-input {
```
Comment (lanwen) [inline war/src/main/js/widgets/add/addform.less]: i'd suggest `job-name_incorrect`

**Comment by samatdav** [comment]
@lanwen I renamed class. Is it fine now?

**Comment by samatdav** [comment]
@michaelneale can you restart the build? It seems like a remote failure.

**Comment by lanwen** [comment]
LGTM

**Comment by daniel-beck** [comment]
When clearing the input field, it retains the previous validation status. Is that deliberate?

**Comment by samatdav** [comment]
@daniel-beck what do you mean by retains? If a recall correctly the behavior is the same as with blur event.

**Comment by michaelneale** [comment]
@samatdav he means that if you delete the contents, it still shows an error around it (which is how I interpret what he said, but I don't think my explanation is any clearer).

**Comment by samatdav** [comment]
@daniel-beck @michaelneale oh, I see - do you mean the red border still shows up after you delete the job name which is invalid? I am actually not sure what happens because the code for replacing warning message and removing warning class name is in one place: https://github.com/jenkinsci/jenkins/pull/2416/files#diff-e146d4fdbfe6385a456aea9775f6282dR253
While the warning text message is removed, the red border is not. May be you could suggest what may be the error and a possibile solution?

**Comment by oleg-nenashev** [comment]
@michaelneale @samatdav I'd guess there is no interest to proceed with this change, right?
If no, I propose to close it.

**Comment by daniel-beck** [comment]
Based on latest review comments, there's also an unresolved issue when clearing the input field.

Closing for now, can always be reopened or resubmitted later.
